frame/staking/src/ledger.rs

```diff
@@ -0,0 +1,375 @@
+	/// Any balance that is becoming free, which may eventually be transferred out of the stash
+	/// (assuming it doesn't get slashed first). It is assumed that this will be treated as a first
+	/// in, first out queue where the new (higher value) eras get pushed on the back.
+	pub unlocking: BoundedVec<UnlockChunk<BalanceOf<T>>, T::MaxUnlockingChunks>,
+	/// List of eras for which the stakers behind a validator have claimed rewards. Only updated
+	/// for validators.
+	pub claimed_rewards: BoundedVec<EraIndex, T::HistoryDepth>,
+	/// The controller associated with this ledger's stash.
+	///
+	/// This is not stored on-chain, and is only bundled when the ledger is read from storage.
+	/// Use [`controller`] function to get the controller associated with the ledger.
+	#[codec(skip)]
+	controller: Option<T::AccountId>,
+}
+
+impl<T: Config> StakingLedger<T> {
+	#[cfg(feature = "runtime-benchmarks")]
+	pub fn default_from(stash: T::AccountId) -> Self {
+		Self {
+			stash,
+			total: Zero::zero(),
+			active: Zero::zero(),
+			unlocking: Default::default(),
+			claimed_rewards: Default::default(),
+			controller: Default::default(),
+		}
+	}
+
+	/// Returns a new instance of a staking ledger.
+	///
+	/// The `Ledger` storage is not mutated. In order to do that [`fn storage_put`] must be called
+	/// on the returned staking ledger.
+	///
+	/// Note: as the controller accounts are being deprecated, the stash account is the same as the
+	/// controller account.
+	pub fn new(
+		stash: T::AccountId,
+		active_stake: BalanceOf<T>,
+		total_stake: BalanceOf<T>,
+		unlocking: BoundedVec<UnlockChunk<BalanceOf<T>>, T::MaxUnlockingChunks>,
+		claimed_rewards: BoundedVec<EraIndex, T::HistoryDepth>,
+	) -> Self {
+		Self {
+			stash: stash.clone(),
+			active: active_stake,
+			total: total_stake,
+			unlocking,
+			claimed_rewards,
+			// controllers are deprecated and map 1-1 to stashes.
+			controller: Some(stash),
+		}
+	}
+
+	/// Returns the controller account of the staking ledger.
+	fn controller(&self) -> T::AccountId {
+		self.controller
+			.clone()
+			.expect("staking ledger instance was fetched through `storage_get`, qed.")
+	}
+
+	/// Returns the staking ledger entry stored in [`Ledger`] storage. Returns `None` if the entry
+	/// does not exist for the give controller.
+	///
+	/// Note: To ensure consistency, all the [`Ledger`] storage queries should be made through this
+	/// helper function.
+	pub(crate) fn storage_get(controller: &T::AccountId) -> Option<Self> {
+		<Ledger<T>>::get(&controller).map(|mut l| {
+			l.controller = Some(controller.clone());
+			l
+		})
+	}
+
+	/// Inserts/updates a staking ledger account.
+	///
+	/// The staking locks od the stash account are updated accordingly.
+	///
+	/// Note: To ensure lock consistency, all the [`Ledger`] storage updates should be made through
+	/// this helper function.
+	pub(crate) fn storage_put(&self) {
+		T::Currency::set_lock(STAKING_ID, &self.stash, self.total, WithdrawReasons::all());
```

Comment:
I wonder if setting a lock here is a bit obscure given the function name. A caller might want to know they are setting a lock on the stash (and not just persisting ledger)?

Reply:
Perhaps `ledger.update()` to be a bit neutral to convey this might be doing more than just put in the storage? Although am not sure about the naming either.

Reply:
Renamed to `ledger.mutate()`